examples/allocation/zig/testdata/greet.zig

```diff
@@ -0,0 +1,65 @@
+const std = @import("std");
+const allocator = std.heap.page_allocator;
+
+// main is required for Zig to compile to enable WASI.
+pub fn main() anyerror!void {
+    return;
+}
+
+extern "env" fn log(ptr: [*]const u8, size: u32) void;
+
+// _log prints a message to the console using log.
+pub fn _log(message: []const u8) void {
+    log(message.ptr, message.len);
+}
+
+pub export fn malloc(length: usize) ?[*]u8 {
```

Comment:
Curious how this optional is translated at the Wasm layer. Returning null just becomes returning 0?

Reply:
> Curious how this optional is translated at the Wasm layer. Returning null just becomes returning 0?

Yes, null becomes `0`. An optional pointer in Zig works the same as a pointer in C, therefore this type matches the C-ABI correctly. Whereas a regular pointer in Zig is much more strict and doesn't allow `null` as a value, meaning you wouldn't be able to convey an allocation failure.

Reply:
that's good. thank you for the confirmation